samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java

```diff
@@ -0,0 +1,498 @@
+
+    private final AppLifeCycleListener appLifeCycleListener;
+
+    TestStreamApplication(String inputTopic, String outputTopic, String processorId,
+        CountDownLatch processedMessagesLatch, AppLifeCycleListener appLifeCycleListener) {
+      this.inputTopic = inputTopic;
+      this.outputTopic = outputTopic;
+      this.processedMessagesLatch = processedMessagesLatch;
+      this.appLifeCycleListener = appLifeCycleListener;
+    }
+
+    @Override
+    public void init(StreamGraph graph, Config config) {
+      try {
+        BiFunction<String, String, String> messageBuilder = (key, msg) -> {
+          String result = "";
+          try {
+            TestKafkaEvent incomingMessage = TestKafkaEvent.fromSerializedString(msg);
+            if (appLifeCycleListener != null) {
+              appLifeCycleListener.onMessageReceived(incomingMessage);
+            }
+            result = incomingMessage.toString();
+          } catch (Exception e) {
+            LOGGER.error("Exception when processing message in TestStreamApplication: ", e);
+            appLifeCycleListener.onExceptionOccurred(e);
+          }
+          if (processedMessagesLatch != null) {
+            processedMessagesLatch.countDown();
+          }
+          return result;
+        };
+        MessageStream<String> inputStream = graph.getInputStream(inputTopic, messageBuilder);
+        OutputStream<String, String, String> outputStream = graph.getOutputStream(outputTopic, event -> null, event -> event);
+        inputStream.sendTo(outputStream);
+      } catch (Exception e) {
+        if (appLifeCycleListener != null) {
+          appLifeCycleListener.onExceptionOccurred(e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void setUp() {
+    super.setUp();
+    testStreamAppName = String.format("test-app-name-%s", UUID.randomUUID().toString());
+    testStreamAppId = String.format("test-app-id-%s", UUID.randomUUID().toString());
+    inputKafkaTopic = String.format("test-input-topic-%s", UUID.randomUUID().toString());
+    outputKafkaTopic = String.format("test-output-topic-%s", UUID.randomUUID().toString());
+    for (String kafkaTopic : ImmutableList.of(inputKafkaTopic, outputKafkaTopic)) {
+      LOGGER.info("Creating kafka topic: {}.", kafkaTopic);
+      TestUtils.createTopic(zkUtils(), kafkaTopic, 5, 1, servers(), new Properties());
+    }
+  }
+
+  @Override
+  public void tearDown() {
+    for (String kafkaTopic : ImmutableList.of(inputKafkaTopic, outputKafkaTopic)) {
+      LOGGER.info("Deleting kafka topic: {}.", kafkaTopic);
+      AdminUtils.deleteTopic(zkUtils(), kafkaTopic);
+    }
+    super.tearDown();
+  }
+
+  // Helper methods that are shared by tests.
+  private List<String> readProcessorIdsFromZK(String zkPathPrefix, List<String> processorIds) {
+    ZkClient zkClient = new ZkClient(zkConnect());
+    ZkKeyBuilder zkKeyBuilder = new ZkKeyBuilder(zkPathPrefix);
+    ZkUtils zkUtils = new ZkUtils(zkKeyBuilder, zkClient, 10000);
```

Comment:
Why not take ZkUtils as a parameter to these helpers instead of creating it every time?